hibernate-core/src/main/java/org/hibernate/sql/ast/spi/SqlExpressionResolver.java

```diff
@@ -90,7 +90,10 @@ static ColumnReferenceKey createColumnReferenceKey(String tableExpression, Selec
 	static ColumnReferenceKey createColumnReferenceKey(TableReference tableReference, SelectableMapping selectable) {
 		assert tableReference.containsAffectedTableName( selectable.getContainingTableExpression() )
 				: String.format( ROOT, "Expecting tables to match between TableReference (%s) and SelectableMapping (%s)", tableReference.getTableId(), selectable.getContainingTableExpression() );
-		return createColumnReferenceKey( tableReference, selectable.getSelectablePath(), selectable.getJdbcMapping() );
+		final JdbcMapping jdbcMapping = (selectable instanceof DiscriminatorMapping) ?
+			((DiscriminatorMapping)selectable).getUnderlyingJdbcMapping() :
+			selectable.getJdbcMapping();
+		return createColumnReferenceKey( tableReference, selectable.getSelectablePath(), jdbcMapping );
```

Comment:
Fixes the discriminator case by using the underlying JDBC type as part of the key to look up the discriminator property in the `sqlSelectionMap` in `DomainResultCreationStateImpl.resolveSqlExpression(ColumnReferenceKey key, Function<SqlAstProcessingState, Expression> creator)` function.